File: 2009.minimum-number-of-operations-to-make-array-continuous.py
#
# @lc app=leetcode id=2009 lang=python3
#
# [2009] Minimum Number of Operations to Make Array Continuous
#

# @lc code=start
from typing import List
class Solution:
    def minOperations(self, nums: List[int]) -> int:

        # all elements in nums must be unique
        # the diff between max and min in nums == len(nums) - 1 

        #constraints 1 <= nums[i] <= 10 ** 9 

        #The key is to find the maximum number of elements that we can keep in the array
        a = sorted(set(nums))
        ans = -1
        left = 0
        for i, x in enumerate(a):
            while a[left] < x - len(nums) + 1:
                left += 1
            ans = max(ans, i - left + 1)
        #how many elements we need to change
        return len(nums) - ans    
# @lc code=end

File: 2009.minimum-number-of-operations-to-make-array-continuous.java

import java.util.*;
// @lc code=start
class Solution {
    public int minOperations(int[] nums) {
        Arrays.sort(nums);

        int n = nums.length;
        int m = 1;
        for(int i = 1; i < n; i++){
            if(nums[i] != nums[i -1]){
                nums[m++] = nums[i]; // remove duplicates in place
            }
        }

        int ans = 0;
        int left = 0;
        for(int i = 0; i < m; i++){
            while(nums[i] - nums[left] > n -1){
                left++;
            }
            ans = Math.max(ans, i - left + 1);
        }
        return n - ans;
    }
}
// @lc code=end

